medcat/cat.py

```diff
@@ -765,20 +764,180 @@ def get_json(self, text, only_cui=False, addl_info=['cui2icd10', 'cui2ontologies
 
         return json.dumps(out)
 
-    def multiprocessing(self, in_data, nproc=8, batch_size_chars=1000000, only_cui=False, addl_info=[]):
-        r''' Run multiprocessing NOT FOR TRAINING
+    def _separate_gpu_multiproc(self):
+        # Loop though the models and check are there GPU devices
+        gpu_components = []
+        for component in self.pipe.nlp.components:
+            if isinstance(component[1], MetaCAT):
+                if component[1].config.general['device'] == 'cuda':
+                    self.pipe.nlp.disable_pipe(component[0])
+                    gpu_components.append(component)
+
+        return gpu_components
+
+    def _run_gpu_components(self, docs, gpu_components, id2text):
+        self.log.debug("Running GPU components separately")
+
+        # First convert the docs into the fake spacy doc format
+        spacy_docs = json_to_fake_spacy(docs, id2text=id2text)
+        # Disable component locks also
+        for name, component in gpu_components:
+            component.config.general['disable_component_lock'] = True
+
+        for name, component in gpu_components:
+            spacy_docs = component.pipe(spacy_docs)
+
+        for spacy_doc in spacy_docs:
+            for ent in spacy_doc.ents:
+                docs[spacy_doc.id]['entities'][ent._.id]['meta_anns'].update(ent._.meta_anns)
+
+        return docs
+
+    def _batch_generator(self, data, batch_size_chars, skip_ids=set()):
+        docs = []
+        char_count = 0
+        for doc in data:
+            if doc[0] not in skip_ids:
+                char_count += len(doc[1])
+                docs.append(doc)
+                if char_count < batch_size_chars:
+                    continue
+                yield docs
+                docs = []
+                char_count = 0
+
+        if len(docs) > 0:
+            yield docs
+
+
+    def _save_docs_to_file(self, docs, annotated_ids, out_split_size, save_dir_path, annotated_ids_path, part_counter=0):
+        path = os.path.join(save_dir_path, 'part_{}.pickle'.format(part_counter))
+        pickle.dump(docs, open(path, "wb"))
+        self.log.info("Saved part: {}, to: {}".format(part_counter, path))
+        part_counter = part_counter + 1 # Increase for save, as it should be what is the next part
+        pickle.dump((annotated_ids, part_counter), open(annotated_ids_path, 'wb'))
+        return part_counter
```

Comment:
Does this mean each call on this will generate a new file on the user-side disk? If true, cannot see any housekeeping atm and how the out-of-space error could be prevented.